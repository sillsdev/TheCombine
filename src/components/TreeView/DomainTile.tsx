import React, { ReactNode } from "react";
import { Button, Grid, Typography } from "@material-ui/core";
import {
  ChevronLeft,
  ChevronRight,
  KeyboardArrowDown,
  KeyboardArrowUp,
} from "@material-ui/icons";
import SemanticDomainWithSubdomains from "../../types/SemanticDomain";

export enum Direction {
  Down,
  Left,
  Right,
  Up,
}

interface DomainTileProps {
  domain: SemanticDomainWithSubdomains;
  onClick: (domain: SemanticDomainWithSubdomains) => any;
  direction?: Direction | undefined;
}

// Creates a semantic domain tile, which can be clicked on to navigate to that semantic domain
export default class DomainTile extends React.Component<DomainTileProps> {
  domainText(domain: SemanticDomainWithSubdomains): ReactNode {
    return (
      <div style={{ textTransform: "capitalize" }}>
        <Typography variant={"overline"}>{domain.id}</Typography>
        <Typography variant={"body1"}>{domain.name}</Typography>
      </div>
    );
  }

  textWithArrow(
    domain: SemanticDomainWithSubdomains,
    direction: Direction | undefined
  ): ReactNode {
    switch (direction) {
      case Direction.Down:
        return (
          <div>
            {this.domainText(domain)}
            <KeyboardArrowDown />
          </div>
        );
      case Direction.Left:
        return (
          <Grid
            container
            alignItems="center"
            justify="space-around"
            wrap="nowrap"
          >
            <Grid item>
              <ChevronLeft />
            </Grid>
            <Grid item>{this.domainText(domain)}</Grid>
          </Grid>
        );
      case Direction.Right:
        return (
          <Grid
            container
            alignItems="center"
            justify="space-around"
            wrap="nowrap"
          >
            <Grid item>{this.domainText(domain)}</Grid>
            <Grid item>
              <ChevronRight />
            </Grid>
          </Grid>
        );
      case Direction.Up:
        return (
          <div>
            <KeyboardArrowUp />
            {this.domainText(domain)}
          </div>
        );
      default:
        return <div>{this.domainText(domain)}</div>;
    }
  }

  render() {
    let domain = this.props.domain;
    return (
      <Button
        id={domain.id}
        color={"primary"}
        variant={"outlined"}
        style={{
          left: 0,
          bottom: 0,
          width: "95%",
          height: "95%",
          margin: "2.5%",
        }}
        onClick={() => {
          this.props.onClick(domain);
        }}
      >
        {this.textWithArrow(domain, this.props.direction)}
      </Button>
    );
  }
}
