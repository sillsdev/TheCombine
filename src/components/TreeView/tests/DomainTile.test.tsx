import React from "react";
import renderer, { ReactTestRenderer } from "react-test-renderer";

import DomainTile, { Direction } from "components/TreeView/DomainTile";
import MockDomain from "components/TreeView/tests/MockSemanticDomain";

var tileMaster: ReactTestRenderer;
const MOCK_ANIMATE = jest.fn();

describe("Tests DomainTile", () => {
  it("Renders directionless (default) tile without crashing", () => {
    createTile();
  });

  it("Renders directional tile without crashing", () => {
    createTile(Direction.Right);
  });

  it("Tile with direction matches the latest snapshot", () => {
    createTile(Direction.Right);
    snapTest();
  });

  it("Click calls function", () => {
    createTile();
    tileMaster.root.findByType("button").props.onClick();
    expect(MOCK_ANIMATE).toHaveBeenCalledTimes(1);
  });
});

// Creates the tile
function createTile(direction?: Direction) {
  renderer.act(() => {
    tileMaster = renderer.create(
      <DomainTile
        domain={MockDomain}
        onClick={MOCK_ANIMATE}
        direction={direction}
      />
    );
  });
}

// Perform a snapshot test
function snapTest() {
  expect(tileMaster.toJSON()).toMatchSnapshot();
}
