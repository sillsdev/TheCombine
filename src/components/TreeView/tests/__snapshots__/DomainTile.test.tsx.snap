// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainTile Tile with direction matches the latest snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
  disabled={false}
  id="1"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "bottom": 0,
      "height": "95%",
      "left": 0,
      "margin": "2.5%",
      "width": "95%",
    }
  }
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-wrap-xs-nowrap MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-space-around"
    >
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <div
          style={
            Object {
              "textTransform": "capitalize",
            }
          }
        >
          <span
            className="MuiTypography-root MuiTypography-overline"
          >
            1
          </span>
          <p
            className="MuiTypography-root MuiTypography-body1"
          >
            parent
          </p>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </div>
    </div>
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</button>
`;
